### PR TITLE
feat: allowVoid option in array-callback-return

### PR DESCRIPTION
**Pull request by Tanujkanti4441**
#### Prerequisites checklist

- [x] I have read the [contributing guidelines](https://github.com/eslint/eslint/blob/HEAD/CONTRIBUTING.md).

#### What is the purpose of this pull request? (put an "X" next to an item)

[ ] Documentation update
[ ] Bug fix ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/bug-report.md))
[ ] New rule ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/rule-proposal.md))
[x] Changes an existing rule ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/rule-change-proposal.md))
[ ] Add autofix to a rule
[ ] Add a CLI option
[ ] Add something to the core
[ ] Other, please explain:

#### What changes did you make? (Give an overview)
Added a new option called `allowVoid` to the `array-callback-return` rule which allow `void` operator to be used in `forEach` callback return. This option will work only with `checkForEach: true` option.

```js
/*eslint array-callback-return: ["error", { checkForEach: true, allowVoid: true }]*/

myArray.forEach(item => void handleItem(item));  // no-error
```
And added related tests

#### Is there anything you'd like reviewers to focus on?

Refs #17285

### CODE REVIEW

**Comment by nopeless** [comment]
Thanks for doing this work 😁

**Comment by Tanujkanti4441** [comment]
> Thanks for doing this work 😁

Any reviews or anything that needs to be added

File: lib/rules/array-callback-return.js [hunk 1]
```diff
@@ -211,6 +215,14 @@ module.exports = {
...
                     messageId = "expectedNoReturnValue";
                 }
+
+                if (options.checkForEach &&
+                    options.allowVoid &&
+                    node.type === "ArrowFunctionExpression" &&
+                    node.body.type === "UnaryExpression" &&
+                    node.body.operator === "void") {
+                    return;
+                }
```
Comment (nzakas) [inline lib/rules/array-callback-return.js:225]: I think we can make this a bit easier to follow by pulling the common checks up a level.

```suggestion
                if (options.checkForEach && node.type === "ArrowFunctionExpression") {

                    if (node.expression) {
                        messageId = "expectedNoReturnValue";
                    }
    
                    if (options.allowVoid &&
                        node.body.type === "UnaryExpression" &&
                        node.body.operator === "void") {
                        return;
                    }
                }
```

**Review by nzakas — CHANGES_REQUESTED** [review]
This looks like a good start! I have a suggestion for refactoring to make things a bit clearer.

Also, please update the documentation to describe the new option and give examples.

File: lib/rules/array-callback-return.js (hunk 1 above)
Comment (Tanujkanti4441) [reply lib/rules/array-callback-return.js]: Done!
I think `node.expression` is `true`  for both the conditions

**Review by mdjermanovic — COMMENTED** [review]
I think this option should apply to `return void`, too:

```js
/*eslint array-callback-return: ["error", { checkForEach: true, allowVoid: true }]*/

myArray.forEach(item => {
    if (item < 0) {
        return void doSomething(item); // should be okay
    }
    doSomethingElse(item);
});

```

Also, the original issue https://github.com/eslint/eslint/issues/17285 mentions implementing [suggestions](https://eslint.org/docs/latest/extend/custom-rules#providing-suggestions) like in https://github.com/eslint/eslint/pull/17282.

File: docs/src/rules/array-callback-return.md [hunk 2]
```diff
@@ -92,10 +92,13 @@ var bar = foo.map(node => node.getAttribute("id"));
...
 ## Options
 
-This rule accepts a configuration object with two options:
+This rule accepts a configuration object with three options:
 
 * `"allowImplicit": false` (default) When set to `true`, allows callbacks of methods that require a return value to implicitly return `undefined` with a `return` statement containing no expression.
 * `"checkForEach": false` (default) When set to `true`, rule will also report `forEach` callbacks that return a value.
+* `"allowVoid": false` (default) When set to `true`, allows `void` in `forEach` callbacks, so rule will not report the return value with a `void` operator.
+
+**Note-** `{ "allowVoid": true }` works only if `checkForEach` option is set to `true`.
```
Comment (nzakas) [inline docs/src/rules/array-callback-return.md:101]:
```suggestion
**Note:** `{ "allowVoid": true }` works only if `checkForEach` option is set to `true`.  
```

**Review by nzakas — APPROVED** [review]
Thanks, this LGTM. Just one suggestion for docs formatting.

Would like another review from the team before merging.

**Comment by nopeless** [comment]
I'm not sure how much of a priority this is, but the `allowVoid` for callback functions have an IDE suggestion as well.

**Review by mdjermanovic — CHANGES_REQUESTED** [review]
I'd be fine with implementing suggestions in a follow-up, but I think the option should apply to `return void` (https://github.com/eslint/eslint/pull/17564#pullrequestreview-1629002364) from the start.

File: tests/lib/rules/array-callback-return.js [hunk 3]
```diff
@@ -217,6 +226,10 @@ ruleTester.run("array-callback-return", rule, {
...
         { code: "foo.filter(function foo() { return; })", options: checkForEachOptions, errors: [{ messageId: "expectedReturnValue", data: { name: "function 'foo'", arrayMethodName: "Array.prototype.filter" } }] },
         { code: "foo.every(cb || function() {})", options: checkForEachOptions, errors: ["Array.prototype.every() expects a return value from function."] },
+        { code: "foo.forEach((x) => void x)", options: checkForEachOptions, parserOptions: { ecmaVersion: 6 }, errors: [{ messageId: "expectedNoReturnValue" }] },
+        { code: "foo.forEach((x) => void bar(x))", options: checkForEachOptions, parserOptions: { ecmaVersion: 6 }, errors: [{ messageId: "expectedNoReturnValue" }] },
+        { code: "foo.forEach((x) => { return void bar(x); })", options: checkForEachOptions, parserOptions: { ecmaVersion: 6 }, errors: [{ messageId: "expectedNoReturnValue" }] },
+        { code: "foo.forEach((x) => { if (a === b) { return void a; } bar(x) })", options: checkForEachOptions, parserOptions: { ecmaVersion: 6 }, errors: [{ messageId: "expectedNoReturnValue" }] },
```
Comment (mdjermanovic) [inline tests/lib/rules/array-callback-return.js:232]: Can you add a few tests to check if `allowVoid: true` allows only `void` expressions? For example:

```suggestion
        { code: "foo.forEach((x) => { if (a === b) { return void a; } bar(x) })", options: checkForEachOptions, parserOptions: { ecmaVersion: 6 }, errors: [{ messageId: "expectedNoReturnValue" }] },
        { code: "foo.forEach((x) => x)", options: checkForEachAllowVoid, parserOptions: { ecmaVersion: 6 }, errors: [{ messageId: "expectedNoReturnValue" }] },
        { code: "foo.forEach((x) => !x)", options: checkForEachAllowVoid, parserOptions: { ecmaVersion: 6 }, errors: [{ messageId: "expectedNoReturnValue" }] },
        { code: "foo.forEach((x) => { return x; })", options: checkForEachAllowVoid, parserOptions: { ecmaVersion: 6 }, errors: [{ messageId: "expectedNoReturnValue" }] },
        { code: "foo.forEach((x) => { return !x; })", options: checkForEachAllowVoid, parserOptions: { ecmaVersion: 6 }, errors: [{ messageId: "expectedNoReturnValue" }] },
```

**Review by mdjermanovic — COMMENTED** [review]
One suggestion about the tests, then LGTM.

File: tests/lib/rules/array-callback-return.js (hunk 3 above)
Comment (Tanujkanti4441) [reply tests/lib/rules/array-callback-return.js:232]: Done!

**Review by mdjermanovic — APPROVED** [review]
LGTM, thanks!

**Pull request merged**